### PR TITLE
Add a restarting check to runAttach

### PR DESCRIPTION
**Pull request by keloyang**
Signed-off-by: yangshukui <yangshukui@huawei.com>

part carried from https://github.com/moby/moby/pull/32137

**- What I did**
Add a restarting check to runAttach

**- How I did it**

**- How to verify it**
terminal 1,
```
docker run -d --restart=always busybox
3664ae34b27306711b55a4cbf1802d1a027b3ab33ef7fe7789c1aefb0268c928
sleep 5
docker attach 3664ae34b27306711b55a4cbf1802d1a027b3ab33ef7fe7789c1aefb0268c928
```
terminal 2,
```
docker stop 3664ae34b27306711b55a4cbf1802d1a027b3ab33ef7fe7789c1aefb0268c928
```
and then docker attach maybe block.

### CODE REVIEW

**Comment by dnephin** [comment]
LGTM once the lint error is fixed. I think if you rebase onto #102 it should be resolved

**Comment by thaJeztah** [comment]
https://github.com/docker/cli/pull/111 was merged, so this now needs a rebase

**Comment by keloyang** [comment]
ping @thaJeztah ,how can fix codecov error ?

**Comment by dnephin** [comment]
The codecov error is saying that this change has 0% test coverage. 

You could write a unit test for `inspectContainerAndCheckState()` using a fake client, which will fix the problem. You can find links to examples of fake clients in #37.

**Comment by thaJeztah** [comment]
@keloyang looks like some changes got lost in your last push.

Also I noticed that your last push changed permissions on one file;

screen shot 2017-06-02 at 16 18 15

**Comment by keloyang** [comment]
@dnephin thanks for your help , I have add some unit testcase.

**Comment by keloyang** [comment]
ping @thaJeztah @dnephin

File: cli/command/container/attach.go [hunk 1]
```diff
@@ -34,11 +34,15 @@ func inspectContainerAndCheckState(ctx context.Context, cli client.APIClient, ar
...
+	if c.State.Restarting {
+		return nil, errors.New("You cannot attach to a restarting container, wait until it is running")
+	}
+
 	return &c, nil
 }
 
 // NewAttachCommand creates a new cobra.Command for `docker attach`
-func NewAttachCommand(dockerCli *command.DockerCli) *cobra.Command {
+func NewAttachCommand(dockerCli command.Cli) *cobra.Command {
```
Comment (thaJeztah) [inline cli/command/container/attach.go]: This change should probably be in the other commit

File: cli/command/container/attach.go [hunk 2]
```diff
@@ -58,7 +62,7 @@ func NewAttachCommand(dockerCli *command.DockerCli) *cobra.Command {
 	return cmd
 }
 
-func runAttach(dockerCli *command.DockerCli, opts *attachOptions) error {
+func runAttach(dockerCli command.Cli, opts *attachOptions) error {
```
Comment (thaJeztah) [inline cli/command/container/attach.go]: This change should probably be in the other commit

File: cli/command/container/attach_test.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+		name                 string
+		args                 []string
+		expectedError        string
+		containerInspectFunc func(img string) (types.ContainerJSON, error)
+	}{
+		{
+			name:          "client-error",
+			args:          []string{"5cb5bb5e4a3b"},
+			expectedError: "something went wrong",
+			containerInspectFunc: func(coantainID string) (types.ContainerJSON, error) {
```
Comment (thaJeztah) [inline cli/command/container/attach_test.go]: nit: typo; `coantainID` (here and in the other tests below

File: cli/command/container/exec_test.go [hunk 4]
```diff
@@ -114,3 +120,31 @@ func compareExecConfig(config1 *types.ExecConfig, config2 *types.ExecConfig) boo
...
+		name                 string
+		args                 []string
+		expectedError        string
+		containerInspectFunc func(img string) (types.ContainerJSON, error)
+	}{
+		{
+			name:          "client-error",
+			args:          []string{"5cb5bb5e4a3b", "-t", "-i", "bash"},
+			expectedError: "something went wrong",
+			containerInspectFunc: func(coantainID string) (types.ContainerJSON, error) {
```
Comment (thaJeztah) [inline cli/command/container/exec_test.go]: nit: here as well; typo `coantainID`

**Comment by keloyang** [comment]
@thaJeztah @dnephin thanks very much ,I have modified my pr.

**Pull request merged**